### PR TITLE
Tell the ciphers which DRBG to use for generating random bytes.

### PR DESCRIPTION
**Pull request by kroeckx**
This changes some of the places that use RAND_bytes() to use ssl specific DRBG.

##### Checklist

- [ ] documentation is added or updated

### CODE REVIEW

File: crypto/evp/evp_locl.h [hunk 1]
```diff
@@ -39,6 +39,7 @@ struct evp_cipher_ctx_st {
     int final_used;
     int block_mask;
     unsigned char final[EVP_MAX_BLOCK_LENGTH]; /* possible final block */
+    RAND_DRBG *drbg;
```
Comment (richsalz) [inline crypto/evp/evp_locl.h]: Does it make sense to not have "just" the DRBG, but the whole SSL object?  I can see the arguments for either way.

File: include/openssl/evp.h [hunk 2]
```diff
@@ -344,6 +344,7 @@ int (*EVP_CIPHER_meth_get_ctrl(const EVP_CIPHER *cipher))(EVP_CIPHER_CTX *,
 # define         EVP_CTRL_SET_PIPELINE_INPUT_BUFS        0x23
 /* Set the input buffer lengths to use for a pipelined operation */
 # define         EVP_CTRL_SET_PIPELINE_INPUT_LENS        0x24
+# define         EVP_CTRL_SET_DRBG                       0x25
```
Comment (levitte) [inline include/openssl/evp.h]: I suggest adding a getter as well, for use with implementations that won't or can't include `evp_locl.h`

Comment (kroeckx) [reply include/openssl/evp.h]: As in an implementation of an EVP_CIPHER_CTX that's not part of OpenSSL? Is that supported by something like an engine?

File: crypto/evp/e_aes_cbc_hmac_sha256.c [hunk 3]
```diff
@@ -170,8 +173,13 @@ static size_t tls1_1_multi_block_encrypt(EVP_AES_HMAC_SHA256 *key,
 #  endif
 
     /* ask for IVs in bulk */
-    if (RAND_bytes((IVs = blocks[0].c), 16 * x4) <= 0)
+    IVs = blocks[0].c;
+    if (drbg != NULL) {
+        if (RAND_DRBG_generate(drbg, IVs, 16 * x4, 0, NULL, 0) <= 0)
```
Comment (mspncp) [inline crypto/evp/e_aes_cbc_hmac_sha256.c]: You could replace RAND_DRBG_generate() by RAND_DRBG_bytes(), now that #4752 has been merged.

File: crypto/evp/e_aes.c [hunk 4]
```diff
@@ -1349,9 +1350,13 @@ static int aes_gcm_ctrl(EVP_CIPHER_CTX *c, int type, int arg, void *ptr)
             return 0;
         if (arg)
             memcpy(gctx->iv, ptr, arg);
-        if (EVP_CIPHER_CTX_encrypting(c)
-            && RAND_bytes(gctx->iv + arg, gctx->ivlen - arg) <= 0)
-            return 0;
+        if (EVP_CIPHER_CTX_encrypting(c)) {
+            if (c->drbg != NULL) {
+                if (RAND_DRBG_generate(c->drbg, gctx->iv + arg, gctx->ivlen - arg, 0, NULL, 0) <= 0)
```
Comment (mspncp) [inline crypto/evp/e_aes.c]: You could replace RAND_DRBG_generate() by RAND_DRBG_bytes(), now that #4752 has been merged.

File: include/openssl/evp.h (hunk 2 above)
Comment (kroeckx) [reply include/openssl/evp.h]: @levitte: I'm waiting for your reply

Comment (levitte) [reply include/openssl/evp.h]: I'm not sure I understand the full implication of your question.  A getter control would retrieve the associated DRBG, regardless of if the content of the `EVP_CIPHER_CTX` is "part of OpenSSL" or not, no?

Comment (kroeckx) [reply include/openssl/evp.h]: The question is if there are cases where it's not part of OpenSSL. I really don't like adding APIs that are never going to get used.

**Comment by kroeckx** [comment]
I've pushed a new commit that should cover more things using an EVP_CIPHER_CTX.

I currently have no idea how I would like to change things like RSA padding. I don't see an easy way to tell those functions which DRBG to use.

File: include/openssl/evp.h (hunk 2 above)
Comment (kroeckx) [reply include/openssl/evp.h]: It's of course about a cipher implementation, not the ctx than can be external. I'll add it.

**Comment by kroeckx** [comment]
I'll look into things like RSA in some other pull request.

**Pull request closed without merging**